src/nodes/hypertable_modify.c

```diff
@@ -93,6 +95,276 @@ get_chunk_dispatch_states(PlanState *substate)
+ * left and right operands, check if one of the operands is
+ * a simple Var type. If its Var type extract its corresponding
+ * column name from hypertable_compression catalogue table.
+ * If extracted column is a SEGMENT BY column then build
+ * scan keys.
+ */
+static bool
+parse_predicate(Node *node, HypertableModifyState *context)
+{
+	if (node == NULL)
+		return false;
+
+	Chunk *curr_chunk = linitial(context->chunks);
+	List *ht_compression_info = ts_hypertable_compression_get(curr_chunk->fd.hypertable_id);
+	Var *var;
+	char *column_name;
+	TypeCacheEntry *tce;
+	int op_strategy;
+
+	switch (nodeTag(node))
+	{
+		case T_OpExpr:
+		{
+			OpExpr *opexpr = (OpExpr *) node;
+			Expr *leftop, *rightop;
+			Const *arg_value;
+
+			leftop = linitial(opexpr->args);
+			rightop = lsecond(opexpr->args);
+
+			if (IsA(leftop, RelabelType))
+				leftop = ((RelabelType *) leftop)->arg;
+			if (IsA(rightop, RelabelType))
+				rightop = ((RelabelType *) rightop)->arg;
+
+			if (IsA(leftop, Var) && IsA(rightop, Const))
+			{
+				var = (Var *) leftop;
+				arg_value = (Const *) rightop;
+			}
+			else if (IsA(rightop, Var) && IsA(leftop, Const))
+			{
+				var = (Var *) rightop;
+				arg_value = (Const *) leftop;
+			}
+			else
+				return false;
+
+			column_name = get_attname(context->chunk_rte->relid, var->varattno, false);
+			if (is_segmentby_col(ht_compression_info, column_name))
+			{
+				tce = lookup_type_cache(var->vartype, TYPECACHE_BTREE_OPFAMILY);
+				op_strategy = get_op_opfamily_strategy(opexpr->opno, tce->btree_opf);
+				if (op_strategy == BTLessStrategyNumber || op_strategy == BTEqualStrategyNumber ||
+					op_strategy == BTGreaterStrategyNumber)
+					build_scankeys(context, column_name, arg_value, op_strategy);
+			}
+		}
+		break;
+		case T_NullTest:
+		{
+			NullTest *ntest = (NullTest *) node;
+			if (IsA(ntest->arg, Var))
+			{
+				int attrno = 0;
+				Var *var = (Var *) ntest->arg;
+				column_name = get_attname(context->chunk_rte->relid, var->varattno, false);
+				if (is_segmentby_col(ht_compression_info, column_name) &&
+					is_segmentby_col_null(context, column_name, &attrno))
+				{
+					context->null_columns = bms_add_member(context->null_columns, attrno);
+				}
+			}
+		}
+		break;
+		default:
+			return false;
+	}
+	return false;
+}
+
+static void
+build_scankeys_for_segmentby_qual(PlanState *ps, HypertableModifyState *context)
+{
+	ListCell *lc;
+	context->chunk_rte = rt_fetch(((Scan *) ps->plan)->scanrelid, ps->state->es_range_table);
+	foreach (lc, ps->plan->qual)
+	{
+		parse_predicate(lfirst(lc), context);
+	}
+}
+
+/*
+ * Traverse the plan tree to look for Scan nodes on uncompressed chunks.
+ * Once Scan node is found check if chunk is compressed, if so then save
+ * the chunk in HypertableModifyState to be used during plan execution.
+ * We also parse the WHERE quals to get information about predicates.
+ */
+static bool
+collect_chunks_from_scan(PlanState *ps, HypertableModifyState *context)
```

Comment:
We do have the list of chunks for the update somewhere, that we evaluate at the planning time, right? Can we use it here? This approach with traversing the plan tree looks unreliable, imagine you have a join with a different hypertable, then you'll find unrelated chunks and everything will break. Something like `explain update ht_metrics_compressed set value = 1 from ht_metrics_compressed t1 join ht_metrics_compressed t2 on t1.value = t2.value + 1;`

Reply:
By the way, a similar query is already broken on main: https://github.com/timescale/timescaledb/issues/5475

Reply:
In short i dont see any issue with this, however iam not sure what can break this.

Long story:

Challenge is that i need to access Filter clauses which are present at Scan node.
As you say we end up saving same chunks in the list and also chunks which represent some other relation specified in the query.
After some testing i see following behaviour w.r.t to below plan:

```
postgres=# explain update sample_table set temperature = 1 from sample_table t1 join sample_table1 t2 on t1.temperature > t2.temperature + 1;
                                                     QUERY PLAN
--------------------------------------------------------------------------------------------------------------------
 Custom Scan (HypertableModify)  (cost=0.26..1246.08 rows=65333 width=38)
   ->  Update on sample_table  (cost=0.26..1246.08 rows=65333 width=38)
         Update on _hyper_9_10_chunk sample_table_1
         ->  Nested Loop  (cost=0.26..1246.08 rows=65333 width=38)
               ->  Nested Loop  (cost=0.26..428.35 rows=9333 width=20)
                     Join Filter: (t1.temperature > (t2.temperature + '1'::double precision))
                     ->  Seq Scan on _hyper_9_10_chunk t1  (cost=0.00..1.07 rows=7 width=18)
                     ->  Custom Scan (DecompressChunk) on _hyper_7_8_chunk t2  (cost=0.26..1.04 rows=4000 width=18)
                           ->  Seq Scan on compress_hyper_8_9_chunk  (cost=0.00..1.04 rows=4 width=136)
               ->  Materialize  (cost=0.00..1.10 rows=7 width=10)
                     ->  Seq Scan on _hyper_9_10_chunk sample_table_1  (cost=0.00..1.07 rows=7 width=10)
(11 rows)
```
In above example i end up having 2 entries in chunk List, which are duplicates. Note Seq Scan on compress_hyper_8_9_chunk will be skipped as its compression status will be set to 0.

So if a chunk on which update/delete is already done, then 2nd call to perform the same operation on same chunk will be a no-op as the qualifying rows have already been moved from compressed chunk to uncompressed chunk.

Reply:
Sven's reply:

ExecQual will always happen after the scan and never be incorporated into the scan, for SeqScan this does not matter as it cannot incorporate scankeys but for index scans it does matter and using ExecQual is not desirable. We need to derive ScanKeys from the quals here as we improve the batch filtering to also work with orderby metadata (in a followup PR)

https://iobeam.slack.com/archives/C032AS21X4P/p1679570721314599?thread_ts=1679563611.989269&cid=C032AS21X4P

Reply:
Well, I would prefer planning time because it's a more "usual" way, and probably it's even simpler? Once for hypertable, you would need to find the conditions that you will use for choosing the segments to decompress. Then, split them into the ScanKey/ExecQual parts (this might be optional at this point, we can start with just ExecQual I guess). Then, for each chunk, you'll need to translate the variables using `ts_make_inh_translation_list()`.

If you feel that redoing this is too much work, we can go with traversing the planner tree at execution time, but please limit it to the immediate children of the ModifyTable node (the ones that show as "Update chunk1" in explain). But I have a hunch that we'll have to redo it in the future anyway.

Reply:
The reason why this has to happen at execution time is to make this work with executor startup exclusion otherwise chunk exclusion will only work with planner chunk exclusion resulting in potentially a lot of unnecessary decompressions. Also I dont think limiting to immediate children is the correct approach here as this has the potential to miss chunks. If we are worried we trigger decompression on other hypertables due to joins we should instead check the chunk belongs to the hypertable we are updating.